tokio/Cargo.toml

```
@@ -104,7 +104,7 @@ memchr = { version = "2.2", optional = true }
 mio = { version = "0.6.20", optional = true }
 iovec = { version = "0.1.4", optional = true }
 num_cpus = { version = "1.8.0", optional = true }
-parking_lot = { version = "0.10.0", optional = true } # Not in full
+parking_lot = { version = "0.11.0", optional = true } # Not in full
```

What do you think about a more permissive requirement that allows both, like
```suggestion
parking_lot = { version = ">= 0.10.0, <= 0.11.0", optional = true } # Not in full
```
If the goal is just to reduce duplicate dependency versions, this would be the ideal choice --- if other dependencies in tree are already pulling 0.10, Cargo would select that version. Tokio doesn't really use any of the APIs that have changed between those releases...

---

Would that prevent using `0.11.1` once that is released?

---

What would be the advantage? Seems best to track deps?

---

> Would that prevent using `0.11.1` once that is released?

Err, I meant `< 0.12.0`.

> What would be the advantage? Seems best to track deps?

The idea is that Tokio should be able to use whatever `parking_lot` version is pulled in by another crate, if there is one in-tree, to minimize compiling duplicate versions. If another crate has an older dependency, we could just use it, and prefer the latest if no other crate depends on some other version.

I'm not sure if this is the right choice or not, but the only `parking_lot` APIs we depend on are the ones that mimic `std`, and therefore will probably not change very much. We can ignore breaking changes to unrelated APIs that we don't care about in order to reduce the total number of `parking_lot`s in a crate where multiple dependencies use `parking_lot`.

This is a loosely-held opinion...I'm willing to be convinced this is a bad idea.

---

I don't have any data on whether or not it works well. If you think it is worth it, we can try it. I also do not know if the cargo dependency resolver is smart enough to dedup the way you are describing (it is a very hard problem).

---

I recall having heard issues with the resolver in cases like this before on the user's forum. I would want to test that it really works properly before doing this.

---

Hmm, not a big deal if this doesn't work, then!

---

Unless something changed within the last ~year, `cargo` will prefer using the latest versions possible for everything even if deduplication would be possible, *unless* the crate being deduplicated uses `links` (out of necessity) - see also <https://github.com/rust-lang/cargo/issues/4902#issuecomment-422038441>.